### PR TITLE
Fixes for p2p-compactblocks.py test timeouts on travis (#8842)

### PR DESCRIPTION
**Pull request by ryanofsky**
Fixes for #8842. There are three commits:

* Commit `[qa] Fix bug in compactblocks v2 merge` fixes a p2p-compactblocks.py timeout related to, but not exactly the same as the timeout reported in #8842.
* Commit `[qa] Fix stale data bug in test_compactblocks_not_at_tip` adds a missing variable initialization that isn't currently a problem, but would lead to test failures after the next commit.
* Commit `Modify getblocktxn handler not to drop requests for old blocks` is the actual fix for the timeout reported in #8842.

### CODE REVIEW

**Comment by laanwj** [comment]
I'm not sure I understand this: is the protocol change needed just to fix the test?

Nit: this should not be labeled "tests" if it also involves a protocol change: if there is a protocol change that would be the primary focus, I was also confused by the title.

**Comment by ryanofsky** [comment]
I'll add more detail to the third commit message. Making nodes always respond to getblocktxn requests instead of silently ignoring requests for old blocks (depth > 10) fixes a race in the syncing protocol that only happens in the test environment, because the test environment generates new blocks very quickly, which wouldn't happen in the real world.

**Comment by TheBlueMatt** [comment]
Note that while this is incredibly unlikely on main net, due to the bursts there, this could conceivably happen on test net.

**Comment by TheBlueMatt** [comment]
Concept ACK

**Comment by ryanofsky** [comment]
Updated the commit message. For completeness, I'll mention that I did also implement an alternate fix for for the timeouts at https://github.com/ryanofsky/bitcoin/commit/36692c9b426a23fb8ce9e32b5ffee7920b31c98b (`Get rid of MAX_GETBLOCKTXN_DEPTH constant`).

File: src/main.cpp [hunk 1]
```diff
@@ -5450,7 +5450,19 @@ bool static ProcessMessage(CNode* pfrom, string strCommand, CDataStream& vRecv,
...
+            // blocktxn response. Sending a full block response instead of a
+            // small blocktxn response is preferable in the case where a peer
+            // might maliciously send lots of getblocktxn requests to trigger
+            // expensive disk reads, because it will require the peer to
+            // actually receive all the data read from disk over the network.
             LogPrint("net", "Peer %d sent us a getblocktxn for a block > %i deep", pfrom->id, MAX_BLOCKTXN_DEPTH);
+            CInv vInv;
+            vInv.type = State(pfrom->GetId())->fWantsCmpctWitness ? MSG_WITNESS_BLOCK : MSG_BLOCK;
+            vInv.hash = req.blockhash;
+            pfrom->vRecvGetData.push_front(vInv);
```
Comment (ryanofsky) [inline src/main.cpp]: Thanks, switched to push_back.

**Comment by sdaftuar** [comment]
utACK dac53b58b555183ccc0d5e64c428528267cd98b3

Note that the behavior this is changing could also be problematic for users on mainnet, say if you had an edge router whose network went down for a couple hours -- on resumption it might relay the first of a batch of blocks as a compact block, but then ignore the subsequent GETBLOCKTXN if its tip rapidly advances in the meantime as it catches up.

Also, I think this has been mentioned elsewhere, but this issue doesn't just affect the p2p-compactblocks.py test, but many other of our python tests as well (as an example, I've been dealing with sporadic failures due to this issue in the pruning test this evening).

**Comment by laanwj** [comment]
Going to merge this, as the random errors in the segwit test are starting to be really annoying.

File: src/main.cpp [hunk 2]
```diff
@@ -5450,7 +5450,19 @@ bool static ProcessMessage(CNode* pfrom, string strCommand, CDataStream& vRecv,
...
         if (it->second->nHeight < chainActive.Height() - MAX_BLOCKTXN_DEPTH) {
+            // If an older block is requested (should never happen in practice,
+            // but can happen in tests) send a block response instead of a
+            // blocktxn response. Sending a full block response instead of a
+            // small blocktxn response is preferable in the case where a peer
+            // might maliciously send lots of getblocktxn requests to trigger
+            // expensive disk reads, because it will require the peer to
+            // actually receive all the data read from disk over the network.
             LogPrint("net", "Peer %d sent us a getblocktxn for a block > %i deep", pfrom->id, MAX_BLOCKTXN_DEPTH);
+            CInv vInv;
```
Comment (TheBlueMatt) [inline src/main.cpp]: Oops, usually v refers to the type being a vector, but this is just a single inv.

**Review by TheBlueMatt — COMMENTED** [review]
Postumous utACK

File: src/main.cpp (hunk 2 above)
Comment (ryanofsky) [inline src/main.cpp]: Thanks, corrected in #9160

**Comment by sdaftuar** [comment]
@TheBlueMatt Should this be tagged for backport to 0.13?

**Comment by TheBlueMatt** [comment]
It's largely a fix for QA instability.... What is our normal process for that?

**Comment by sdaftuar** [comment]
Well usually QA instability is addressed by changing the tests, and I think we tend to be pretty liberal about backporting test changes.

Personally I think it makes sense to backport this behavior change, as I mentioned above I think this is basically a bugfix since it could be triggered in some mainnet configurations as well, and I think the patch here is small and low risk.  Not sure if everyone would agree with my assessment though...?

**Comment by TheBlueMatt** [comment]
Yea, I'm OK with this being backported.

**Pull request merged**